proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,252 @@
+# Interface instructions
+
+This proposal defines new instructions that operate over interface value types. These instructions have the same form as core instructions; they push and pop values on an execution stack and support single pass validation.
+
+The new interface instructions are summarized below.
+
+```
+#interface-instr ::=
+  | bool.lift_iMM
+  | bool.lower_iMM
+
+  | sNN.lift_iMM
+  | sNN.lower_iMM
+
+  | uNN.lift_iMM
+  | uNN.lower_iMM
+
+  | string.lift_memory ..
+  | string.lower_memory ..
+
+  | record.lift ..
+  | record.lower ..
+
+  | array.lift_memory ..
+  | array.lower_memory ..
+
+  | invoke ..
+
+where:
+  NN is one of [8, 16, 32, 64]
+  MM is one of [32, 64]
+```
+
+In addition the following instructions are included from the core instructions:
+
+```
+#interface-instr ::=
+  | ..
+  | <numeric instructions>
+  | <memory instructions>
+  | <table instructions>
+  | drop
+  | unreachable
+  | nop
+```
+
+Refer to the definition of instruction classes [here](https://webassembly.github.io/spec/core/syntax/instructions.html).
+
+## Laziness
+
+Interface values generated by interface instructions are lazy. The execution of an instruction generating an interface value happens immediately before the instruction that consumes the value.
+
+## `bool.lift_iMM`
+
+This instruction has type `[iMM] -> [bool]`.
+
+The output is `true` iff the input `iMM` has any bit set in the integer.
+
+## `bool.lower_iMM`
+
+This instruction has type `[bool] -> iMM`.
+
+The output is `0b1` zero extended to the desired bitwidth iff the input is true.
+
+## `sNN.lift_iMM`
+
+This instruction has type `iMM` -> `sNN`.
+
+The output is the value of `iMM` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `sNN.lower_iMM`
+
+This instruction has type `[sNN] -> [iMM]`
+
+The output is the value of `sNN` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lift_iMM`
+
+This instruction has type `iMM` -> `uNN`.
+
+The output is the value of `iMM` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
```

Comment:
need two variants of these instructions: those that can cause errors and those that do not (by ignoring errors)
also, trapping is very user-hostile for this.
Suggest adding a branch to the error-reporting variants; i.e., break on failure. (This requires that we at least import the normal block instructions into IT adapters.)

Reply:
Is it possible for us to import the block instructions if we're not adding control flow instructions?

I agree that trapping is very user-hostile, I left this behavior as a temporary design as we don't have block instructions.

Reply:
we will need block instructions. At the least we will need them for exception processing.
We should be careful about which block instructions we import; i.e., on a case-by-case basis.

Reply:
I've renamed references from trapping to 'failing' and added a section on 'failure' that specifies it is currently trapping but will be replaced with something better later.